tests/unit/cluster.tcl

```diff
@@ -222,4 +221,80 @@ start_server [list overrides $base_conf] {
 }
 }
 
-} ;# tags
\ No newline at end of file
+# Test redis-cli --cluster create, add-node.
+# Test that one slot can be migrated to and then away from the new node.
+start_server [list overrides $base_conf] {
+start_server [list overrides $base_conf] {
+start_server [list overrides $base_conf] {
+start_server [list overrides $base_conf] {
+
+    test {Migrate the last slot away from a node} {
+        # Create a cluster of 3 nodes
+        exec src/redis-cli --cluster-yes --cluster create \
+                           127.0.0.1:[srv 0 port] \
+                           127.0.0.1:[srv -1 port] \
+                           127.0.0.1:[srv -2 port]
+
+        wait_for_condition 1000 50 {
+            [csi 0 cluster_state] eq {ok} &&
+            [csi -1 cluster_state] eq {ok} &&
+            [csi -2 cluster_state] eq {ok}
+        } else {
+            fail "Cluster doesn't stabilize"
+        }
+
+        # Insert some data
+        assert_equal OK [exec src/redis-cli -c -p [srv 0 port] SET foo bar]
+        set slot [exec src/redis-cli -c -p [srv 0 port] CLUSTER KEYSLOT foo]
+
+        # Add new node to the cluster
+        exec src/redis-cli --cluster-yes --cluster add-node \
+                     127.0.0.1:[srv -3 port] \
+                     127.0.0.1:[srv 0 port]
+
+        wait_for_condition 1000 50 {
+            [csi 0 cluster_state] eq {ok} &&
+            [csi -1 cluster_state] eq {ok} &&
+            [csi -2 cluster_state] eq {ok} &&
+            [csi -3 cluster_state] eq {ok}
+        } else {
+            fail "Cluster doesn't stabilize"
+        }
+
+        set newnode_r [redis_client -3]
+        set newnode_id [$newnode_r CLUSTER MYID]
+
+        # Find out which node has the key "foo" by asking the new node for a
+        # redirect.
+        catch { $newnode_r get foo } e
+        assert_match "MOVED $slot *" $e
+        set owner [lindex $e 2]
+        set owner_id [exec src/redis-cli -u redis://$owner CLUSTER MYID]
+
+        # Move slot to new node using plain Redis commands
+        assert_equal OK [$newnode_r CLUSTER SETSLOT $slot IMPORTING $owner_id]
+        assert_equal OK [exec src/redis-cli -u redis://$owner \
+                             CLUSTER SETSLOT $slot MIGRATING $newnode_id]
+        assert_equal {foo} [exec src/redis-cli -u redis://$owner \
+                                CLUSTER GETKEYSINSLOT $slot 10]
+        assert_equal OK [exec src/redis-cli -u redis://$owner \
+                             MIGRATE 127.0.0.1 [srv -3 port] "" 0 5000 KEYS foo]
+        assert_equal OK [$newnode_r CLUSTER SETSLOT $slot NODE $newnode_id]
+        assert_equal OK [exec src/redis-cli -u redis://$owner \
+                             CLUSTER SETSLOT $slot NODE $newnode_id]
+
+        # Move the only slot back to original node
+        exec src/redis-cli --cluster reshard 127.0.0.1:[srv -3 port] \
```

Comment:
this test is not guaranteed to catch the original race 100% but short of adding test hooks to redis-cli (to intentionally hold off calling setslot on the source) I guess this is the best that can be done.

Reply:
That's right.

But I think we can reproduce the condition using plain Redis commands. It will be a test with a different purpose: to check that the new owner pingpongs the source node and that the source turns itself into a replica. I can add such test case.

Reply:
I added a check that the node turns itself into a replica but I didn't repeat the complete test to verify that you get that "ERR Please use SETSLOT only with masters" in this scenario. (I know that it happens sometimes with the current test though. I tried without the fix to ignore the error.)

I might turn this test case into a loop with two alternatives (cli reshard and manual migrate).

Reply:
LGTM. I love the refactoring of start_multiple_servers :-)